Zend/zend_ini.c

```diff
@@ -536,6 +541,202 @@ ZEND_API bool zend_ini_parse_bool(zend_string *str)
+
+	zend_ulong retval;
+	errno = 0;
+
+	if (signed_result) {
+		retval = (zend_ulong) ZEND_STRTOL(digits, &digits_end, 0);
+	} else {
+		retval = ZEND_STRTOUL(digits, &digits_end, 0);
+	}
+
+	if (errno == ERANGE) {
+		overflow = true;
+	} else if (!signed_result) {
+		/* ZEND_STRTOUL() does not report a range error when the subject starts
+		 * with a minus sign, so we check this here. Ignore "-1" as it is
+		 * commonly used as max value, for instance in memory_limit=-1. */
+		if (digits[0] == '-' && !(digits_end - digits == 2 && digits_end == str_end && digits[1] == '1')) {
+			overflow = true;
+		}
+	}
+
+	if (UNEXPECTED(digits_end == digits)) {
+		/* No leading digits */
+
+		/* Escape the string to avoid null bytes and to make non-printable chars
+		 * visible */
+		smart_str_append_escaped(&invalid, ZSTR_VAL(value), ZSTR_LEN(value));
+		smart_str_0(&invalid);
+
+		*errstr = zend_strpprintf(0, "Invalid quantity \"%s\": no valid leading digits, interpreting as \"0\" for backwards compatibility",
+						ZSTR_VAL(invalid.s));
+
+		smart_str_free(&invalid);
+		return 0;
+	}
+
+	/* Allow for whitespace between integer portion and any suffix character */
+	while (digits_end < str_end && zend_is_whitespace(*digits_end)) ++digits_end;
+
+	/* No exponent suffix. */
+	if (digits_end == str_end) {
+		goto end;
+	}
+
+	switch (*(str_end-1)) {
+		case 'g':
+		case 'G':
+			factor = 1<<30;
+			break;
+		case 'm':
+		case 'M':
+			factor = 1<<20;
+			break;
+		case 'k':
+		case 'K':
+			factor = 1<<10;
+			break;
+		default:
+			/* Unknown suffix */
+			smart_str_append_escaped(&invalid, ZSTR_VAL(value), ZSTR_LEN(value));
+			smart_str_0(&invalid);
+			smart_str_append_escaped(&interpreted, str, digits_end - str);
+			smart_str_0(&interpreted);
+			smart_str_append_escaped(&chr, str_end-1, 1);
+			smart_str_0(&chr);
+
+			*errstr = zend_strpprintf(0, "Invalid quantity \"%s\": unknown multipler \"%s\", interpreting as \"%s\" for backwards compatibility",
+						ZSTR_VAL(invalid.s), ZSTR_VAL(chr.s), ZSTR_VAL(interpreted.s));
+
+			smart_str_free(&invalid);
+			smart_str_free(&interpreted);
+			smart_str_free(&chr);
+
+			return retval;
+	}
+
+	if (!overflow) {
+		if (signed_result) {
+			zend_long sretval = (zend_long)retval;
+			if (sretval > 0) {
+				overflow = (zend_long)retval > ZEND_LONG_MAX / (zend_long)factor;
+			} else {
+				overflow = (zend_long)retval < ZEND_LONG_MIN / (zend_long)factor;
+			}
+		} else {
+			overflow = retval > ZEND_ULONG_MAX / factor;
+		}
+	}
+
+	retval *= factor;
+
+	if (UNEXPECTED(digits_end != str_end-1)) {
+		/* More than one character in suffix */
+		smart_str_append_escaped(&invalid, ZSTR_VAL(value), ZSTR_LEN(value));
+		smart_str_0(&invalid);
+		smart_str_append_escaped(&interpreted, str, digits_end - str);
+		smart_str_0(&interpreted);
+		smart_str_append_escaped(&chr, str_end-1, 1);
+		smart_str_0(&chr);
+
+		*errstr = zend_strpprintf(0, "Invalid quantity \"%s\", interpreting as \"%s%s\" for backwards compatibility",
```

Comment:
If there's an overflow from `ZEND_STRTOL` **and** invalid quantity, you'll only get the error message for the latter. Probably unlikely though, as this could only happen when literally entering very large numbers.

Reply:
This can also happen for the "invalid multiplier case" since overflow can happen in strtol().

I decided to emit only one error (or to mention only one problem in the error) for the sake of simplicity. Do you think that it would be worth it to change this ?

Reply:
Nah, it's fine.